### PR TITLE
[enhancement] enable bf16 BLAS level-3 support

### PR DESCRIPTION
**Pull request by icfaust**
# Description
Add gpu_blas_compute_mode kwarg to set_config and config_context. This will enable bf16 support when gpu_blas_compute_mode is set to a proper string which matches from the [oneMKL compute_mode enum list](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-dpcpp/2024-0/compute-modes.html).

Changes proposed in this pull request:
- Add ComputeMode enum via pybind11 from mkl_dal_sycl.hpp when dpc backend is available
- Set and/or acquire os environment variable "DAL_BLAS_COMPUTE_MODE" when dpc backend is available via set_config/config_context kwarg gpu_blas_compute_mode

tasks:
- [ ] verify float_to_bf16 operation

### CODE REVIEW

**Review by samir-nasibli — COMMENTED** [review]
Please cover this functionality with some tests.
Probably you can extend this test suit https://github.com/intel/scikit-learn-intelex/blob/master/sklearnex/tests/test_config.py

**Comment by icfaust** [comment]
> Please cover this functionality with some tests. Probably you can extend this test suit https://github.com/intel/scikit-learn-intelex/blob/master/sklearnex/tests/test_config.py

Definitely a good idea. I'll work on it

File: sklearnex/linear_model/tests/test_linear.py [hunk 1]
```diff
@@ -75,3 +75,30 @@ def test_sklearnex_import_elastic():
...
+
+    size = 100
+    X = np.ones((size, 2), dtype=np.float32)
+    y = np.arange(size, dtype=np.float32) * np.finfo(np.float32).eps
+    X[:, 0] = y
+    X = _convert_to_dataframe(X, sycl_queue=queue, target_df=dataframe)
+    y = _convert_to_dataframe(y, sycl_queue=queue, target_df=dataframe)
+    linreg_standard = LinearRegression().fit(X, y)
+    with config_context(compute_mode="float_to_bf16"):
+        linreg_bf16 = LinearRegression().fit(X, y)
```
Comment (samir-nasibli) [inline sklearnex/linear_model/tests/test_linear.py:96]: You can offload without dpnp/dpctl dataframes, you can use config_context for that.
```suggestion
    with config_context(target_offload="gpu:0"):
        linreg_standard = LinearRegression().fit(X, y)
    with config_context(compute_mode="float_to_bf16", target_offload="gpu:0"):
        linreg_bf16 = LinearRegression().fit(X, y)
```

File: sklearnex/_config.py [hunk 2]
```diff
@@ -63,6 +75,11 @@ def set_config(target_offload=None, allow_fallback_to_host=None, **sklearn_confi
         If True, allows to fallback computation to host device
         in case particular estimator does not support the selected one.
         Global default: False.
+    compute_mode : str or list, default=None
+        The computational method used for BLAS level 3 routines in
+        oneDAL for select GPU devices. The compute_mode string(s)
+        must be from the oneMKL compute_mode enum list.
```
Comment (Alexsandruss) [inline sklearnex/_config.py:81]: `compute_mode` is too broad term for config parameter. `mkl_blas_compute_mode` might be better.

**Comment by samir-nasibli** [comment]
@icfaust please run intel ci for GPU validation as well

File: sklearnex/linear_model/tests/test_linear.py [hunk 3]
```diff
@@ -75,3 +76,54 @@ def test_sklearnex_import_elastic():
...
+    assert linreg_standard.n_features_in_ == 2
+    assert linreg_bf16.n_features_in_ == 2
+    assert_raises(
+        AssertionError,
+        assert_allclose,
+        _as_numpy(linreg_standard.coef_),
+        _as_numpy(linreg_bf16.coef_),
+    )
+    print(linreg_bf16.coef_)
+    print(linreg_bf16.intercept_)
```
Comment (samir-nasibli) [inline sklearnex/linear_model/tests/test_linear.py:128]: Please remove prints from tests

**Comment by napetrov** [comment]
But with current approach with env variable we can control this without python by just setting env variable without changes for scikit. Do we need any form of python support here?

File: sklearnex/_config.py [hunk 4]
```diff
@@ -76,6 +96,28 @@ def set_config(target_offload=None, allow_fallback_to_host=None, **sklearn_confi
...
+                )
+                gpu_blas_compute_mode = ",".join(
+                    [ComputeMode.__members__[i.lower()].name for i in inp]
+                )
+                local_config["gpu_blas_compute_mode"] = gpu_blas_compute_mode
+                os.environ["DAL_BLAS_COMPUTE_MODE"] = gpu_blas_compute_mode.upper()
+
+            except (KeyError, AttributeError) as e:
+                raise ValueError(
+                    f"'{e.args[0]}' is not a supported gpu_blas_compute_mode"
```
Comment (KulikovNikita) [inline sklearnex/_config.py:115]: 1) Looks too compolex to be a single function or an isolated piece of code
2) I would recommend to actually use a map:
```python
@lru_cache
def get_compute_mode_map():
    members = ComputeMode.__members__ 
    return { m.name.lower() : m for m in members }

def get_raw_compute_mode():
    ...

def set_raw_compute_mode(cm):
    ...
```

File: sklearnex/linear_model/tests/test_linear.py [hunk 5]
```diff
@@ -81,3 +82,46 @@ def test_sklearnex_import_elastic():
     assert "daal4py" in elasticnet.__module__
     assert_allclose(elasticnet.intercept_, 1.451, atol=1e-3)
     assert_allclose(elasticnet.coef_, [18.838, 64.559], atol=1e-3)
+
```
Comment (KulikovNikita) [inline sklearnex/linear_model/tests/test_linear.py:85]: I would actually recommend to also check for such small numbers in an actual algorithm implementation in order to not get errors while inverting values

File: sklearnex/linear_model/tests/test_linear.py [hunk 6]
```diff
@@ -81,3 +82,46 @@ def test_sklearnex_import_elastic():
...
+    with config_context(
+        target_offload="gpu:0",
+        allow_fallback_to_host=False,
+        gpu_blas_compute_mode="standard",
+    ):
+        linreg_standard = LinearRegression().fit(X, y)
+
+    with config_context(
+        target_offload="gpu:0",
```
Comment (KulikovNikita) [inline sklearnex/linear_model/tests/test_linear.py:110]: Should we use `gpu:0` or rather just `gpu`?

File: sklearnex/linear_model/tests/test_linear.py [hunk 7]
```diff
@@ -81,3 +82,46 @@ def test_sklearnex_import_elastic():
...
+        allow_fallback_to_host=False,
+        gpu_blas_compute_mode="float_to_bf16",
+    ):
+        linreg_bf16 = LinearRegression().fit(X, y)
+
+    assert linreg_standard.n_features_in_ == 2
+    assert linreg_bf16.n_features_in_ == 2
+    # DAL_BLAS_COMPUTE_MODE is available in the mkl_gpu_fpk in 2024.0.2
+    if daal_check_version((2024, "P", 100)):
```
Comment (KulikovNikita) [inline sklearnex/linear_model/tests/test_linear.py:119]: What about NOT testing it in previous versions at all?

File: sklearnex/tests/test_config.py [hunk 8]
```diff
@@ -37,3 +44,54 @@ def test_set_config_works():
...
+        pytest.param(
+            ["FLOAT_TO_BF16", "float_to_bf16x2", "float_to_bf16x3"], marks=gpu_mark
+        ),
+        pytest.param(
+            "float_to_bf16,float_to_bf16x2,float_to_bf16x3",
+            marks=gpu_mark,
+        ),
+        pytest.param("any", marks=gpu_mark),
+    ],
+)
```
Comment (KulikovNikita) [inline sklearnex/tests/test_config.py:69]: Please consider creating separate wrapper

**Comment by KulikovNikita** [comment]
These changes looks very promising!

File: sklearnex/linear_model/tests/test_linear.py (hunk 7 above)
Comment (icfaust) [reply sklearnex/linear_model/tests/test_linear.py:119]: I was hoping to verify operation with and without bf16 enablement. This seemed to me to be the best way of having my cake and eating it too.

File: sklearnex/linear_model/tests/test_linear.py (hunk 6 above)
Comment (icfaust) [reply sklearnex/linear_model/tests/test_linear.py:110]: After this CI run, I will try it.

File: sklearnex/linear_model/tests/test_linear.py (hunk 7 above)
Comment (samir-nasibli) [reply sklearnex/linear_model/tests/test_linear.py:119]: Could you please clarify why do you need to test it for not bf16 enablement?

File: sklearnex/tests/test_config.py (hunk 8 above)
Comment (samir-nasibli) [inline sklearnex/tests/test_config.py:69]: Make sure that you have correct logs/names for tests.
[setting2] is not informative [link](https://dev.azure.com/daal/daal4py/_build/results?buildId=30631&view=logs&j=517fe804-fa30-5dc2-1413-330699242c05&t=6200c9a9-fd17-566a-4b28-c9cb31ff00c5&l=1359)

```
s/sklearnex/tests/test_config.py::test_set_gpu_blas_compute_mode[standard] PASSED [ 18%]
s/sklearnex/tests/test_config.py::test_set_gpu_blas_compute_mode[FORCE_ALTERNATE] PASSED [ 19%]
s/sklearnex/tests/test_config.py::test_set_gpu_blas_compute_mode[setting2] PASSED [ 19%]
s/sklearnex/tests/test_config.py::test_set_gpu_blas_compute_mode[float_to_bf16,float_to_bf16x2,float_to_bf16x3] PASSED  [ 19%]
s/sklearnex/tests/test_config.py::test_set_gpu_blas_compute_mode[any] PASSED [ 20%]
```

**Comment by icfaust** [comment]
I will be closing this PR after a set of verifications

**Pull request closed without merging**